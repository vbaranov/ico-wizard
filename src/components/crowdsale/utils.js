import { noContractAlert } from '../../utils/alerts'
import {
  getCurrentAccount,
  attachToSpecificCrowdsaleContract,
} from '../../utils/blockchainHelpers'
import { contractStore, crowdsalePageStore, tokenStore, web3Store } from '../../stores'
import { toJS } from 'mobx'
import { BigNumber } from 'bignumber.js'

BigNumber.config({ DECIMAL_PLACES : 18 })

export const toBigNumber = (value) => isNaN(value) || value === '' ? new BigNumber(0) : new BigNumber(value)

export let getTokenData = (initCrowdsaleContract, execID, account) => {
  return new Promise((resolve, reject) => {

    let registryStorageObj = toJS(contractStore.registryStorage)

    let getTokenName = initCrowdsaleContract.methods.name(registryStorageObj.addr, execID).call();
    let getTokenSymbol = initCrowdsaleContract.methods.symbol(registryStorageObj.addr, execID).call();
    let getTokenDecimals = initCrowdsaleContract.methods.decimals(registryStorageObj.addr, execID).call();
    let getTokenTotalSupply = initCrowdsaleContract.methods.totalSupply(registryStorageObj.addr, execID).call();
    let getBalanceOf = initCrowdsaleContract.methods.balanceOf(registryStorageObj.addr, execID, account).call();

    return Promise.all([
      getTokenName,
      getTokenSymbol,
      getTokenDecimals,
      getTokenTotalSupply,
      getBalanceOf
    ])
      .then(([
          name,
          ticker,
          decimals,
          totalSupply,
          balanceOf
      ]) => {
        const { web3 } = web3Store

        tokenStore.setProperty('name', web3.utils.toAscii(name))
        console.log('token name: ' + name)
        tokenStore.setProperty('ticker', web3.utils.toAscii(ticker))
        console.log('token ticker: ' + ticker)
        tokenStore.setProperty('decimals', decimals)
        console.log('token decimals: ' + decimals)
        totalSupply = totalSupply * toBigNumber(10).pow(Number(decimals))
        tokenStore.setProperty('supply', totalSupply)
        console.log('token supply: ' + totalSupply)
        balanceOf = Number(balanceOf).toFixed()
        console.log('balanceOf: ' + balanceOf)
        const tokenAmountOf = crowdsalePageStore.tokenAmountOf ? crowdsalePageStore.tokenAmountOf : 0
        console.log('tokenAmountOf: ' + tokenAmountOf)
        crowdsalePageStore.setProperty('tokenAmountOf', tokenAmountOf + parseInt(balanceOf, 10))

        resolve()
      })
      .catch(reject)
  })
}

export let getCrowdsaleData = (initCrowdsaleContract, execID, account, isMintedCappedCrowdsale) => {
  return new Promise((resolve, reject) => {
    if (!initCrowdsaleContract) {
      noContractAlert()
      reject('no contract')
    }

    console.log(initCrowdsaleContract)

    let registryStorageObj = toJS(contractStore.registryStorage)

    let getCrowdsaleInfo = initCrowdsaleContract.methods.getCrowdsaleInfo(registryStorageObj.addr, execID).call();
    let getCurrentTierInfo = initCrowdsaleContract.methods.getCurrentTierInfo(registryStorageObj.addr, execID).call();
    let getTokensSold = initCrowdsaleContract.methods.getTokensSold(registryStorageObj.addr, execID).call();
    let getContributors = isMintedCappedCrowdsale ? initCrowdsaleContract.methods.getCrowdsaleUniqueBuyers(registryStorageObj.addr, execID).call() : null;
    let getCrowdsaleMaxRaise = initCrowdsaleContract.methods.getCrowdsaleMaxRaise(registryStorageObj.addr, execID).call();

    return Promise.all([
      getCrowdsaleInfo,
      getCurrentTierInfo,
      getTokensSold,
      getContributors,
      getCrowdsaleMaxRaise
    ])
      .then(([
          crowdsaleInfo,
          currentTierInfo,
          tokensSold,
          contributors,
          crowdsaleMaxRaise
      ]) => {
        const { web3 } = web3Store

        console.log('crowdsaleInfo:')
        console.log(crowdsaleInfo)
        console.log('currentTierInfo:')
        console.log(currentTierInfo)
        console.log('tokensSold:')
        console.log(tokensSold)
        console.log('contributors:')
        console.log(contributors)
        console.log('crowdsaleMaxRaise:')
        console.log(crowdsaleMaxRaise)

        crowdsalePageStore.setProperty('weiRaised', Number(crowdsaleInfo.wei_raised).toFixed())
        crowdsalePageStore.setProperty('ethRaised', web3.utils.fromWei(crowdsalePageStore.weiRaised, 'ether'))
        crowdsalePageStore.setProperty('rate', Number(currentTierInfo.tier_price).toFixed()) //should be one token in wei
        const storedTokensSold = toBigNumber(tokensSold)
        crowdsalePageStore.setProperty('tokensSold', storedTokensSold.toFixed())
        if (contributors)
          crowdsalePageStore.setProperty('investors', toBigNumber(contributors).toFixed())

        crowdsalePageStore.setProperty('maximumSellableTokens', toBigNumber(crowdsaleMaxRaise.total_sell_cap).toFixed())
        crowdsalePageStore.setProperty('maximumSellableTokensInWei', toBigNumber(crowdsaleMaxRaise.wei_raise_cap).toFixed())

        resolve()
      })
      .catch(reject)
  })
}

export function initializeAccumulativeData() {
  crowdsalePageStore.setProperty('tokenAmountOf', 0)
  return Promise.resolve()
}

export let getCrowdsaleTargetDates = (initCrowdsaleContract, execID) => {
  return new Promise((resolve, reject) => {
    if (!initCrowdsaleContract) {
      noContractAlert()
      reject('no contract')
    }

    console.log(initCrowdsaleContract)

    let registryStorageObj = toJS(contractStore.registryStorage)

    let getCrowdsaleStartAndEndTimes = initCrowdsaleContract.methods.getCrowdsaleStartAndEndTimes(registryStorageObj.addr, execID).call();
    let getCurrentTierInfo = initCrowdsaleContract.methods.getCurrentTierInfo(registryStorageObj.addr, execID).call();

    return Promise.all([getCrowdsaleStartAndEndTimes, getCurrentTierInfo])
      .then(([crowdsaleStartAndEndTimes, currentTierInfo]) => {
        let crowdsaleStartTime = crowdsaleStartAndEndTimes.start_time
        console.log("crowdsaleStartTime:", crowdsaleStartTime)
        const startsAtMilliseconds = crowdsaleStartTime * 1000
        console.log('currentTierInfo:')
        console.log(currentTierInfo)
        let crowdsaleEndTime = currentTierInfo.tier_ends_at
        const endsAtMilliseconds = crowdsaleEndTime * 1000

        crowdsalePageStore.addTier({
          startDate: startsAtMilliseconds,
          endDate: endsAtMilliseconds
        })

        if (!crowdsalePageStore.startDate || crowdsalePageStore.startDate > startsAtMilliseconds)
          crowdsalePageStore.startDate = startsAtMilliseconds
        console.log('startDate:' + startsAtMilliseconds)

        if (!crowdsalePageStore.endDate || crowdsalePageStore.endDate < endsAtMilliseconds)
          crowdsalePageStore.setProperty('endDate', endsAtMilliseconds)
        console.log("endDate:", endsAtMilliseconds)

        resolve()
      })
      .catch(reject)
  })
}

export let isFinalized = (initCrowdsaleContract, crowdsaleExecID) => {
  let registryStorageObj = toJS(contractStore.registryStorage)
  let getCrowdsaleInfo = initCrowdsaleContract.methods.getCrowdsaleInfo(registryStorageObj.addr, crowdsaleExecID).call();

  return getCrowdsaleInfo.then(crowdsaleInfo => {
    let isFinalized = crowdsaleInfo.is_finalized;
    return isFinalized;
  })
}

export const getTiers = () => {
  return getCurrentAccount()
    .then(account => {
      return attachToSpecificCrowdsaleContract("initCrowdsale")
        .then((initCrowdsaleContract) => {
          const { methods } = initCrowdsaleContract
          let registryStorageObj = toJS(contractStore.registryStorage)

          return methods.getCrowdsaleTierList(registryStorageObj.addr, contractStore.crowdsale.execID).call()
            .then(tiers => {
              console.log("tiers:", tiers)
              return Promise.resolve(tiers.length)
            })
        })
    })
}

export const getContractStoreProperty = (contract, property) => {
  const text = contractStore && contractStore[contract] && contractStore[contract][property]
  return text === undefined ? '' : text
}
