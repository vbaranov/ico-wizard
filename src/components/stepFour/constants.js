export const alertOptions = { time: 10000, position: 'top right' }
export const fileDownloadedToasterMsg = "A file with contracts and metadata downloaded on your computer"