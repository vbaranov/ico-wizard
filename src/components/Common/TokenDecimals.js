import React from 'react'
import { composeValidators, isLessOrEqualThan, isNonNegative, isRequired } from '../../utils/validations'
import { TEXT_FIELDS } from '../../utils/constants'
import { Field } from 'react-final-form'
import { InputField2 } from './InputField2'
import { acceptPositiveIntegerOnly } from '../../utils/utils'

export const TokenDecimals = ({ disabled, errorStyle }) => (
  <Field
    validate={composeValidators(
      isRequired(),
      isNonNegative(),
      isLessOrEqualThan("Should not be greater than 18")(18)
    )}
    component={InputField2}
    parse={acceptPositiveIntegerOnly}
    side="left"
    name="decimals"
    type="text"
    description="Refers to how divisible a token can be, from 0 (not at all divisible) to 18 (pretty much continuous)."
    label={TEXT_FIELDS.DECIMALS}
    disabled={disabled}
    errorStyle={errorStyle}
  />
)
