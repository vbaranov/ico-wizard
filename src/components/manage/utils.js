import { isObservableArray } from 'mobx'
import {
  getCurrentAccount,
  sendTXToContract,
  methodToExec
} from '../../utils/blockchainHelpers'
import { contractStore, crowdsaleStore, generalStore, tierStore, tokenStore, web3Store } from '../../stores'
import { VALIDATION_TYPES } from '../../utils/constants'
import { toFixed } from '../../utils/utils'
import { toBigNumber } from '../crowdsale/utils'
import { generateContext } from '../stepFour/utils'
import { BigNumber } from 'bignumber.js'

const { VALID } = VALIDATION_TYPES

const formatDate = timestamp => {
  const ten = i => (i < 10 ? '0' : '') + i
  const date = new Date(timestamp * 1000)
  const YYYY = date.getFullYear()
  const MM = ten(date.getMonth() + 1)
  const DD = ten(date.getDate())
  const HH = ten(date.getHours())
  const II = ten(date.getMinutes())

  return YYYY + '-' + MM + '-' + DD + 'T' + HH + ':' + II
}

export const updateTierAttribute = (attribute, value, tierIndex) => {
  let methodInterface
  let getParams
  const { decimals } = tokenStore
  let methods = {
    //startTime: 'setStartsAt', // startTime is not changed from migration to Auth_os
    endTime: 'updateTierDuration',
    //supply: 'setMaximumSellableTokens', // supply is not changed from migration to Auth_os
    //rate: 'updateRate', // rate is not changed from migration to Auth_os
    whitelist: 'whitelistMultiForTier'
  }

  if (attribute === 'startTime' || attribute === 'endTime' || attribute === 'supply' || attribute === 'whitelist') {
    /*if (attribute === 'startTime') {
      value = toFixed(parseInt(Date.parse(value) / 1000, 10).toString())
    } else */

    if (attribute === 'endTime') {
      let { startTime, endTime } = tierStore.tiers[tierIndex]
      console.log(startTime, endTime)
      const duration = new Date(endTime) - new Date(startTime)
      const durationBN = (toBigNumber(duration)/1000).toFixed()
      value = durationBN
      methodInterface = ["uint256","uint256","bytes"]
      getParams = updateDurationParams
    } /*else if (attribute === 'supply') {
      value = toBigNumber(value).times(`1e${tokenStore.decimals}`).toFixed()
    } */else if (attribute === 'whitelist')  {
      // whitelist
      const rate = tierStore.tiers[tierIndex].rate;
      const rateBN = new BigNumber(rate)
      const oneTokenInETH = rateBN.pow(-1).toFixed()
      const oneTokenInWEI = web3Store.web3.utils.toWei(oneTokenInETH, 'ether')
      value = value.reduce((toAdd, whitelist) => {
        toAdd[0].push(whitelist.addr)
        toAdd[1].push(toBigNumber(whitelist.min).times(oneTokenInWEI).toFixed())
        toAdd[2].push(toBigNumber(whitelist.max).times(oneTokenInWEI).toFixed())
        //toAdd[1].push(whitelist.min * 10 ** decimals ? toFixed((whitelist.min * 10 ** decimals).toString()) : 0)
        //toAdd[2].push(whitelist.max * 10 ** decimals ? toFixed((whitelist.max * 10 ** decimals).toString()) : 0)
        return toAdd
      }, [[], [], []])
      methodInterface = ["uint256","address[]","uint256[]","uint256[]","bytes"]
      getParams = updateWhitelistParams
    }
  }

  /*if (attribute === 'rate') {
    const oneTokenInETH = floorToDecimals(TRUNC_TO_DECIMALS.DECIMALS18, 1 / Number(value))
    value = web3Store.web3.utils.toWei(oneTokenInETH, 'ether')
  }*/

  console.log("value:", value)

  console.log("attribute:", attribute)
  console.log("methods[attribute]:", methods[attribute])

  console.log("tierIndex:", tierIndex)
  console.log("value:", value)

  const paramsToExec = [ tierIndex, value, methodInterface ] // tierIndex + 1 due to `The index of the tier whose duration will be updated (indexes in the tier list are 1-indexed: 0 is an invalid index)`

  const method = methodToExec(`${methods[attribute]}(${methodInterface.join(',')})`, "crowdsaleConsole", getParams, paramsToExec)

  return getCurrentAccount()
    .then(account => {
      const opts = { gasPrice: generalStore.gasPrice, from: account }
      return method.estimateGas(opts)
      .then(estimatedGas => {
        opts.gasLimit = estimatedGas
        return sendTXToContract(method.send(opts))
      })
    })
}

const updateDurationParams = (tierIndex, duration, methodInterface) => {
  console.log(tierIndex, duration)
  const { web3 } = web3Store
  let context = generateContext(0);
  let encodedParameters = web3.eth.abi.encodeParameters(methodInterface, [tierIndex, duration, context]);
  return encodedParameters;
}

const updateWhitelistParams = (tierIndex, [addr, min, max], methodInterface) => {
  console.log(tierIndex, addr, min, max, methodInterface)
  const { web3 } = web3Store
  let context = generateContext(0);
  let encodedParameters = web3.eth.abi.encodeParameters(methodInterface, [tierIndex, addr, min, max, context]);
  return encodedParameters;
}

const crowdsaleData = (tier, crowdsale, token) => {
  const { web3 } = web3Store
  let startsAt = tier.tier_start
  let endsAt = tier.tier_end
  let rate = tier.tier_price
  let tokenName = web3.utils.toAscii(token.token_name)
  let tokenSymbol = web3.utils.toAscii(token.token_symbol)
  let decimals = token.token_decimals
  let multisigWallet = crowdsale.team_wallet
  let tierName = web3.utils.toAscii(tier.tier_name)
  let maximumSellableTokens = tier.tier_sell_cap
  let isUpdatable = tier.duration_is_modifiable
  let isWhitelisted = tier.whitelist_enabled
  let isFinalized = crowdsale.is_finalized
  let whitelistAccounts = tier.whitelist

  return Promise.all([
    multisigWallet,
    startsAt,
    endsAt,
    rate,
    isUpdatable,
    isWhitelisted,
    maximumSellableTokens,
    isFinalized,
    tierName,
    whitelistAccounts,
    [tokenName, tokenSymbol, decimals]
  ]);
}

export const processTier = (tier, crowdsale, token, tierNum) => {
  console.log("tier:", tier)
  console.log("crowdsale:", crowdsale)
  console.log("token:", token)
  const { web3 } = web3Store

  const newTier = {
    whitelist: []
  }

  const initialValues = {}

  return crowdsaleData(tier, crowdsale, token)
    .then(([
             walletAddress,
             startsAt,
             endsAt,
             rate,
             updatable,
             isWhitelisted,
             maximumSellableTokens,
             isFinalized,
             name,
             whitelistAccounts,
             [tokenName, tokenSymbol, decimals]
           ]) => {
      crowdsaleStore.setSelectedProperty('finalized', isFinalized)
      crowdsaleStore.setSelectedProperty('updatable', crowdsaleStore.selected.updatable || updatable)

      newTier.walletAddress = walletAddress
      newTier.startTime = formatDate(startsAt)
      newTier.endTime = formatDate(endsAt)
      newTier.updatable = updatable
      newTier.tier = name

      initialValues.updatable = newTier.updatable
      initialValues.index = tierNum
      initialValues.addresses = {
        crowdsaleAddress: contractStore.crowdsale.execID
      }

      if (tierNum === 0) {
        newTier.whitelistEnabled = isWhitelisted ? 'yes' : 'no'
      }

      return Promise.all([maximumSellableTokens, whitelistAccounts, rate, [tokenName, tokenSymbol, decimals]])
    })
    .then(([maximumSellableTokens, whitelistAccounts, rate, [tokenName, tokenSymbol, decimals]]) => {
      tokenStore.setProperty('name', tokenName)
      tokenStore.setProperty('ticker', tokenSymbol)
      tokenStore.setProperty('decimals', decimals)

      //total supply
      const tokenDecimals = !isNaN(decimals) ? decimals : 0
      const maxCapBeforeDecimals = toBigNumber(maximumSellableTokens).div(`1e${tokenDecimals}`)

      newTier.supply = maxCapBeforeDecimals ? maxCapBeforeDecimals.toFixed() : 0

      return Promise.all([whitelistAccounts, rate, tokenDecimals])
    })
    .then(([whitelistAccounts, rate, tokenDecimals]) => {
      //const { decimals } = tokenStore
      //const tokenDecimals = !isNaN(decimals) ? decimals : 0

      //price
      newTier.rate = toBigNumber(web3.utils.fromWei(toBigNumber(rate).toFixed(), 'ether'))
        .pow(-1)
        .decimalPlaces(0)
        .toFixed()

      tierStore.addTier(newTier, {
        tier: VALID,
        walletAddress: VALID,
        rate: VALID,
        supply: VALID,
        startTime: VALID,
        endTime: VALID,
        updatable: VALID
      })

      const whitelist = newTier.whitelist.slice()

      if (whitelistAccounts) {
        whitelistAccounts.forEach(({ addr, min, max }) => {
          min = parseInt(toFixed(min), 10) * newTier.rate / 10 ** tokenDecimals
          max = parseInt(toFixed(max), 10) * newTier.rate / 10 ** tokenDecimals

          whitelist.push({ addr, min, max, stored: true })
        })
      }

      tierStore.setTierProperty(whitelist, 'whitelist', tierNum)
      tierStore.sortWhitelist(tierNum)

      if (initialValues.updatable) {
        initialValues.startTime = newTier.startTime
        initialValues.endTime = newTier.endTime
        initialValues.rate = newTier.rate
        initialValues.supply = newTier.supply
        initialValues.whitelist = whitelist
      }
      crowdsaleStore.addInitialTierValues(initialValues)
    })
}

export function getFieldsToUpdate(updatableTiers, tiers) {
  const keys = Object
    .keys(updatableTiers[0])
    .filter(key => key !== 'index' && key !== 'updatable' && key !== 'addresses')

  const toUpdate = updatableTiers
    .reduce((toUpdate, tier, index) => {
      keys.forEach(key => {
        let newValue = tiers[tier.index][key]

        if (isObservableArray(newValue)) {
          newValue = newValue.filter(item => !item.stored)

          if (newValue.length) {
            toUpdate.push({ key, newValue, tier: index })
          }

        } else if (newValue !== tier[key]) {
          toUpdate.push({ key, newValue, tier: index })
        }
      })
      return toUpdate
    }, [])
    .sort((item1, item2) => {
      if (item1.tier !== item2.tier) {
        return item2.tier - item1.tier
      }

      if (item1.key === 'startTime' && item2.key === 'endTime') {
        return 1
      }

      if (item1.key === 'endTime' && item2.key === 'startTime') {
        return -1
      }

      return 0
    })

  return toUpdate
}
