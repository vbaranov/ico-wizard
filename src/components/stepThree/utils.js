import moment from 'moment';
import { StepThreeFormMintedCapped } from './StepThreeFormMintedCapped'
import { StepThreeFormDutchAuction } from './StepThreeFormDutchAuction'
import { CROWDSALE_STRATEGIES } from '../../utils/constants'

export function defaultCompanyStartDate() {
  let crowdsaleStartDate = moment().add(5, 'minutes');
  let crowdsaleStartDateFormatted = crowdsaleStartDate.format('YYYY-MM-DDTHH:mm');
  return crowdsaleStartDateFormatted;
}

export const defaultCompanyEndDate = (startDate) => {
  const crowdsaleEndDate = moment(startDate).add(4, 'days').startOf('day')
  return crowdsaleEndDate.format('YYYY-MM-DDTHH:mm')
}

export const getStep3Component = (strategy) => {
  switch(strategy) {
    case CROWDSALE_STRATEGIES.MINTED_CAPPED_CROWDSALE:
      return StepThreeFormMintedCapped;
    case CROWDSALE_STRATEGIES.DUTCH_AUCTION:
      return StepThreeFormDutchAuction;
    default:
      return StepThreeFormMintedCapped;
  }
}
