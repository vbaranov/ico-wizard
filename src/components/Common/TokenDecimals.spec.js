import React from 'react'
import { TokenDecimals } from './TokenDecimals'
import { Form } from 'react-final-form'
import Adapter from 'enzyme-adapter-react-15'
import { configure, mount, shallow } from 'enzyme'
import { VALIDATION_MESSAGES } from '../../utils/constants'

configure({ adapter: new Adapter() })

describe('TokenDecimals', () => {
  it(`should render TokenDecimals component`, () => {
    const wrapper = shallow(
      <Form onSubmit={jest.fn()} component={TokenDecimals} errorStyle={{ color: 'red', fontWeight: 'bold', }}/>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it(`should render TokenDecimals component and its children`, () => {
    const wrapper = mount(
      <Form onSubmit={jest.fn()} component={TokenDecimals} errorStyle={{ color: 'red', fontWeight: 'bold', }}/>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it(`should render TokenDecimals component and its children, with input field disabled`, () => {
    const wrapper = mount(
      <Form
        onSubmit={jest.fn()}
        component={TokenDecimals}
        errorStyle={{ color: 'red', fontWeight: 'bold', }}
        disabled={true}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it(`should give error if decimals is empty`, () => {
    const wrapper = mount(
      <Form onSubmit={jest.fn()} component={TokenDecimals} errorStyle={{ color: 'red', fontWeight: 'bold', }}/>
    )

    const input = wrapper.find('input[name="decimals"]')

    input.simulate('change', { target: { value: '10' } })
    expect(wrapper.find('InputField2').prop('meta').error).toBeFalsy()

    input.simulate('change', { target: { value: '' } })
    expect(wrapper.find('InputField2').prop('meta').error).toBe(VALIDATION_MESSAGES.REQUIRED)
  })

  it(`should give error if decimals is greater than 18`, () => {
    const wrapper = mount(
      <Form onSubmit={jest.fn()} component={TokenDecimals} errorStyle={{ color: 'red', fontWeight: 'bold', }}/>
    )
    const input = wrapper.find('input[name="decimals"]')
    input.simulate('change', { target: { value: '21' } })

    expect(wrapper.find('.error').text()).toBe('Should not be greater than 18')
  })
})
