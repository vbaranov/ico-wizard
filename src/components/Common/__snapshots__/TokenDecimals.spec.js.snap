// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenDecimals should render TokenDecimals component 1`] = `
<TokenDecimals
  __versions={
    Object {
      "final-form": "4.2.0",
      "react-final-form": "3.1.4",
    }
  }
  batch={[Function]}
  blur={[Function]}
  change={[Function]}
  dirty={false}
  dirtySinceLastSubmit={false}
  errorStyle={
    Object {
      "color": "red",
      "fontWeight": "bold",
    }
  }
  errors={Object {}}
  focus={[Function]}
  handleSubmit={[Function]}
  initialize={[Function]}
  invalid={false}
  mutators={Object {}}
  pristine={true}
  reset={[Function]}
  submitFailed={false}
  submitSucceeded={false}
  submitting={false}
  touched={Object {}}
  valid={true}
  validating={false}
  values={Object {}}
  visited={Object {}}
/>
`;

exports[`TokenDecimals should render TokenDecimals component and its children 1`] = `
<ReactFinalForm(4.2.0)(3.1.4)
  component={[Function]}
  errorStyle={
    Object {
      "color": "red",
      "fontWeight": "bold",
    }
  }
  onSubmit={[Function]}
>
  <TokenDecimals
    __versions={
      Object {
        "final-form": "4.2.0",
        "react-final-form": "3.1.4",
      }
    }
    batch={[Function]}
    blur={[Function]}
    change={[Function]}
    dirty={false}
    dirtySinceLastSubmit={false}
    errorStyle={
      Object {
        "color": "red",
        "fontWeight": "bold",
      }
    }
    errors={Object {}}
    focus={[Function]}
    handleSubmit={[Function]}
    initialize={[Function]}
    invalid={false}
    mutators={Object {}}
    pristine={true}
    reset={[Function]}
    submitFailed={false}
    submitSucceeded={false}
    submitting={false}
    touched={Object {}}
    valid={true}
    validating={false}
    values={Object {}}
    visited={Object {}}
  >
    <Field
      component={[Function]}
      description="Refers to how divisible a token can be, from 0 (not at all divisible) to 18 (pretty much continuous)."
      errorStyle={
        Object {
          "color": "red",
          "fontWeight": "bold",
        }
      }
      format={[Function]}
      label="Decimals"
      name="decimals"
      parse={[Function]}
      side="left"
      type="text"
      validate={[Function]}
    >
      <InputField2
        description="Refers to how divisible a token can be, from 0 (not at all divisible) to 18 (pretty much continuous)."
        errorStyle={
          Object {
            "color": "red",
            "fontWeight": "bold",
          }
        }
        input={
          Object {
            "name": "decimals",
            "onBlur": [Function],
            "onChange": [Function],
            "onFocus": [Function],
            "value": "",
          }
        }
        label="Decimals"
        meta={
          Object {
            "active": false,
            "data": Object {},
            "dirty": false,
            "dirtySinceLastSubmit": false,
            "error": "This field is required",
            "initial": undefined,
            "invalid": true,
            "length": undefined,
            "pristine": true,
            "submitError": undefined,
            "submitFailed": false,
            "submitSucceeded": false,
            "touched": false,
            "valid": false,
            "visited": false,
          }
        }
        side="left"
        type="text"
      >
        <div
          className="left"
        >
          <label
            className="label"
            htmlFor="decimals"
          >
            Decimals
          </label>
          <input
            autoComplete="off"
            className="input"
            id="decimals"
            name="decimals"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="text"
            value=""
          />
          <p
            className="description"
          >
            Refers to how divisible a token can be, from 0 (not at all divisible) to 18 (pretty much continuous).
          </p>
          <Error
            errorStyle={
              Object {
                "color": "red",
                "fontWeight": "bold",
              }
            }
            name="decimals"
          >
            <Field
              format={[Function]}
              name="decimals"
              parse={[Function]}
              render={[Function]}
              subscription={
                Object {
                  "error": true,
                  "pristine": true,
                  "touched": true,
                }
              }
            >
              <span>
                <p
                  className="error"
                  key="0"
                  style={
                    Object {
                      "color": "red",
                      "fontWeight": "bold",
                    }
                  }
                />
              </span>
            </Field>
          </Error>
        </div>
      </InputField2>
    </Field>
  </TokenDecimals>
</ReactFinalForm(4.2.0)(3.1.4)>
`;

exports[`TokenDecimals should render TokenDecimals component and its children, with input field disabled 1`] = `
<ReactFinalForm(4.2.0)(3.1.4)
  component={[Function]}
  disabled={true}
  errorStyle={
    Object {
      "color": "red",
      "fontWeight": "bold",
    }
  }
  onSubmit={[Function]}
>
  <TokenDecimals
    __versions={
      Object {
        "final-form": "4.2.0",
        "react-final-form": "3.1.4",
      }
    }
    batch={[Function]}
    blur={[Function]}
    change={[Function]}
    dirty={false}
    dirtySinceLastSubmit={false}
    disabled={true}
    errorStyle={
      Object {
        "color": "red",
        "fontWeight": "bold",
      }
    }
    errors={Object {}}
    focus={[Function]}
    handleSubmit={[Function]}
    initialize={[Function]}
    invalid={false}
    mutators={Object {}}
    pristine={true}
    reset={[Function]}
    submitFailed={false}
    submitSucceeded={false}
    submitting={false}
    touched={Object {}}
    valid={true}
    validating={false}
    values={Object {}}
    visited={Object {}}
  >
    <Field
      component={[Function]}
      description="Refers to how divisible a token can be, from 0 (not at all divisible) to 18 (pretty much continuous)."
      disabled={true}
      errorStyle={
        Object {
          "color": "red",
          "fontWeight": "bold",
        }
      }
      format={[Function]}
      label="Decimals"
      name="decimals"
      parse={[Function]}
      side="left"
      type="text"
      validate={[Function]}
    >
      <InputField2
        description="Refers to how divisible a token can be, from 0 (not at all divisible) to 18 (pretty much continuous)."
        disabled={true}
        errorStyle={
          Object {
            "color": "red",
            "fontWeight": "bold",
          }
        }
        input={
          Object {
            "name": "decimals",
            "onBlur": [Function],
            "onChange": [Function],
            "onFocus": [Function],
            "value": "",
          }
        }
        label="Decimals"
        meta={
          Object {
            "active": false,
            "data": Object {},
            "dirty": false,
            "dirtySinceLastSubmit": false,
            "error": "This field is required",
            "initial": undefined,
            "invalid": true,
            "length": undefined,
            "pristine": true,
            "submitError": undefined,
            "submitFailed": false,
            "submitSucceeded": false,
            "touched": false,
            "valid": false,
            "visited": false,
          }
        }
        side="left"
        type="text"
      >
        <div
          className="left"
        >
          <label
            className="label"
            htmlFor="decimals"
          >
            Decimals
          </label>
          <input
            autoComplete="off"
            className="input"
            disabled={true}
            id="decimals"
            name="decimals"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="text"
            value=""
          />
          <p
            className="description"
          >
            Refers to how divisible a token can be, from 0 (not at all divisible) to 18 (pretty much continuous).
          </p>
          <Error
            errorStyle={
              Object {
                "color": "red",
                "fontWeight": "bold",
              }
            }
            name="decimals"
          >
            <Field
              format={[Function]}
              name="decimals"
              parse={[Function]}
              render={[Function]}
              subscription={
                Object {
                  "error": true,
                  "pristine": true,
                  "touched": true,
                }
              }
            >
              <span>
                <p
                  className="error"
                  key="0"
                  style={
                    Object {
                      "color": "red",
                      "fontWeight": "bold",
                    }
                  }
                />
              </span>
            </Field>
          </Error>
        </div>
      </InputField2>
    </Field>
  </TokenDecimals>
</ReactFinalForm(4.2.0)(3.1.4)>
`;
